src/FakeItEasy/Expressions/ExpressionArgumentConstraintFactory.cs

```diff
@@ -72,18 +72,14 @@ private static IArgumentConstraint CreateEqualityConstraint(object expressionVal
             return new EqualityArgumentConstraint(expressionValue);
         }
 
-        private static object InvokeExpression(Expression expression)
-        {
-            return Expression.Lambda(expression).Compile().DynamicInvoke();
-        }
-
         private IArgumentConstraint GetArgumentConstraintFromExpression(Expression expression)
         {
             object expressionValue = null;
 
             var trappedConstraints = this.argumentConstraintTrapper.TrapConstraints(() =>
             {
-                expressionValue = InvokeExpression(expression);
+                // in this ParsedArgumentExpression scenario, we get benefits from the same optimizations as ParsedCallExpression
+                expressionValue = ParsedCallExpression.FastEvaluate(expression);

```

Comment:
> Having it hidden away and hard to find doesn't seem a bad place to start!

It should be hidden from the public API (which it is), but not hidden from internal code... It should be in a place where we can find it when we need it, and `ParsedCallExpression` is not where I would look for it. I think the best place for `FastEvaluate` would be as an extension method in `ExpressionExtensions`.

> Such other spots in FakeItEasy's code:

Yes, there are definitely more places where this could be useful